### PR TITLE
Adding a unit test to verify if CacheTagHelper gets disabled when setting the enabled flag to false.

### PR DESCRIPTION
**Pull request by sornaks**
/cc: @dougbu @NTaylorMullen

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 1]
```diff
@@ -320,6 +318,47 @@ public void GenerateKey_WithMultipleVaryByOptions_CreatesCombinedKey()
         }
 
         [Fact]
+        public async Task ProcessAsync_DoesNotUseCache_OnPassingFalseWhenEnabled()
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: `ProcessAsync_DoesNotUseCache_OnPassingFalseWhenEnabled` => `ProcessAsync_DoesNotUseCache_WhenEnabledFalse`

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 2]
```diff
@@ -320,6 +318,47 @@ public void GenerateKey_WithMultipleVaryByOptions_CreatesCombinedKey()
...
+            {
+                ViewContext = GetViewContext(),
+                MemoryCache = cache.Object,
+                Enabled = true
+            };
+
+            // Act
+            await cacheTagHelper.ProcessAsync(tagHelperContext, tagHelperOutput);
+
+            // set enabled to false in the cache tag helper.
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: self explanatory, no need for comment.

**Comment by NTaylorMullen** [comment]
:shipit:

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 3]
```diff
@@ -320,6 +318,47 @@ public void GenerateKey_WithMultipleVaryByOptions_CreatesCombinedKey()
...
         [Fact]
+        public async Task ProcessAsync_DoesNotUseCache_OnPassingFalseWhenEnabled()
+        {
+            // Arrange
+            var cache = new Mock<IMemoryCache>();
+            cache.Setup(c => c.Set(
+                /*key*/ It.IsAny<string>(),
+                /*link*/ It.IsAny<IEntryLink>(),
+                /*state*/ It.IsAny<object>(),
+                /*create*/ It.IsAny<Func<ICacheSetContext, object>>()))
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: nit: mock setups in MVC tend to use trailing comments e.g. `It.IsAny<object>(),    // state`

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 4]
```diff
@@ -283,7 +282,6 @@ public void GenerateKey_WithMultipleVaryByOptions_CreatesCombinedKey()
             var id = "unique-id";
             var childContent = "original-child-content";
             var cache = new Mock<IMemoryCache>();
-            cache.CallBase = true;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: why are these lines changing?

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 5]
```diff
@@ -320,6 +318,47 @@ public void GenerateKey_WithMultipleVaryByOptions_CreatesCombinedKey()
...
+            await cacheTagHelper.ProcessAsync(tagHelperContext, tagHelperOutput);
+
+            // Assert
+            // if cache was not disabled this would have been called more than once.
+            cache.Verify(c => c.Set(
+                /*key*/ It.IsAny<string>(),
+                /*link*/ It.IsAny<IEntryLink>(),
+                /*state*/ It.IsAny<object>(),
+                /*create*/ It.IsAny<Func<ICacheSetContext, object>>()),
+                Times.Once);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: why is this method called at all when caching is disabled?

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 6]
```diff
@@ -320,6 +318,47 @@ public void GenerateKey_WithMultipleVaryByOptions_CreatesCombinedKey()
...
+            cache.Setup(c => c.Set(
+                /*key*/ It.IsAny<string>(),
+                /*link*/ It.IsAny<IEntryLink>(),
+                /*state*/ It.IsAny<object>(),
+                /*create*/ It.IsAny<Func<ICacheSetContext, object>>()))
+                .Returns(new DefaultTagHelperContent().SetContent("ok"))
+                .Verifiable();
+            object cacheResult;
+            cache.Setup(c => c.TryGetValue(It.IsAny<string>(), It.IsAny<IEntryLink>(), out cacheResult))
+                .Returns(false);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: does test fail without this setup?  expected cache to be completely unused when tag helper is disabled.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs (hunk 5 above)
Comment (sornaks) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: The test case is about having it enabled initially and then making a call with disabled.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs (hunk 4 above)
Comment (sornaks) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: CallBase is not needed. It is an interface..

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs (hunk 5 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: It's called once for the first `ProcessAsync`

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 7]
```diff
@@ -320,6 +318,47 @@ public void GenerateKey_WithMultipleVaryByOptions_CreatesCombinedKey()
...
+            var tagHelperContext = GetTagHelperContext("unique-id", "original-child-content");
+            var tagHelperOutput = new TagHelperOutput("cache", new Dictionary<string, object>());
+            var cacheTagHelper = new CacheTagHelper
+            {
+                ViewContext = GetViewContext(),
+                MemoryCache = cache.Object,
+                Enabled = true
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: OIC.  this is a bizarre test.  highly unlikely `Enabled` will change after initialization, never after `ProcessAsync` has run.

Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: Hmm right. Is it even possible for a tag helper's property to change after being initialized? (In any sane way.)

Comment (sornaks) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: I dont think it is. So I was trying to test a scenario where one cache tag helper had some value in the cache and a successive cache tag helper tried to access the cache with `enabled=false` in which case the cache must not be used. I am going to write a functional test for this instead (without any time bounds).

**Comment by dougbu** [comment]
> I am going to write a functional test for this instead

@sornaks should this PR be closed now?

**Comment by sornaks** [comment]
Yeah. I'll send that in a separate PR.

**Pull request closed without merging**